File: lecture_07/primes.py
"""
A simple python program to test CPU performance by
generating a series of prime numbers
"""

NUM_PRIMES = 100_000

def is_prime(number):
    "Returns True if 'number' is prime"
    limit = int(number ** 0.5) + 1
    for d in range(2, limit):
        if number % d == 0:
            return False
    return True

def gen_primes(num_primes):
    "Generates first 'num' series of prime numbers"
    v = 2
    while num_primes:
        if is_prime(v):
            print(v)
            num_primes -= 1
        v += 1

if __name__ == '__main__':
    from time import time
    import sys

    print(f"Generating {NUM_PRIMES} prime numbers...", file=sys.stderr)
    start_time = time()
    gen_primes(NUM_PRIMES)
    duration = time() - start_time
    print(f"gen_primes({NUM_PRIMES}) took {duration} seconds", file=sys.stderr)

File: lecture_03/primes.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <math.h>

#include "stop_watch.h"

#define NUM_PRIMES 100000

bool is_prime(unsigned int number) {
    unsigned int limit = sqrt(number) + 1;

    for (unsigned int d=2; d < limit; d++) {
        if (number % d == 0)
            return false;
    }
    return true;
}

void gen_primes(unsigned int num_primes)
{

    for (unsigned int v=2; num_primes > 0; v++) {
        if (is_prime(v)) {
            printf("%u\n", v);
            --num_primes;
        }
    }
}

int main()
{
    stop_watch_t watch;

    fprintf(stderr, "Generating first %u prime numbers...\n", NUM_PRIMES);

    start_watch(&watch);
    gen_primes(NUM_PRIMES);
    stop_watch(&watch);

    fprintf(stderr, "gen_primes(%u) took %3.6Lf seconds.\n",
                    NUM_PRIMES, duration(&watch));

    return EXIT_SUCCESS;
}
